graphql-language/src/parser.rs

```diff
@@ -0,0 +1,363 @@
+#![allow(dead_code, unused_variables)]
+use ast::*;
+use lexer::Punctuator::*;
+use lexer::Token::{
+  Eof,
+  Punctuator,
+  Name as TokenName,
+  IntValue as TokenInt,
+  FloatValue as TokenFloat,
+  StringValue as TokenString,
+};
+use lexer::{Lexer, Token};
+use source::Source;
+use std::iter::Peekable;
+
+pub type Res<'a, T> = Result<T, &'a str>;
+pub type DRes<'a, T> = Res<'a, (T, usize)>;
+
+pub struct Parser<'a> {
+  lexer: Peekable<Lexer<'a>>,
+  source: Source<'a>,
+}
+
+impl<'a> Parser<'a> {
+  pub fn new(src: Source<'a>) -> Self {
```

Comment:
I know I've said differently in the past, but an iterable is just to clunky here. We are working with heavily nested types, not a stream of information. Prefer a function here like `parse_source`, that function could even run it's own lexer so the signature would be `pub fn parse_source(source: Source<'a>) -> Document<'a>`.

The only argument I can see for keeping the parser is the `loc` function, but I don't think that is very compelling argument at all.